File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    public static void main(String[] args) {
      
    }

    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    public static Optional<String> firstThenLowerCase(List<String> list, Predicate<String> predicate) {
        return list.stream()
                .filter(predicate)
                .findFirst()
                .map(String::toLowerCase);
    }

    static record Sayer(String phrase) {
        Sayer and(String word) {
            String separator = phrase.endsWith(" ") || word.startsWith(" ") ? "" : " ";
            return new Sayer(phrase + separator + word);
        }
    }

    public static Sayer say() {
        return new Sayer("");
    }

    public static Sayer say(String word) {
        return new Sayer(word);
    }

    public static long meaningfulLineCount(String filePath) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            return reader.lines()
                    .filter(line -> !line.trim().isEmpty())
                    .count();
        }
    }
}

record Quaternion(double a, double b, double c, double d) {

    static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    static final Quaternion I = new Quaternion(0, 1, 0, 0);
    static final Quaternion J = new Quaternion(0, 0, 1, 0);
    static final Quaternion K = new Quaternion(0, 0, 0, 1);

    Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    Quaternion plus(Quaternion other) {
        return new Quaternion(a + other.a, b + other.b, c + other.c, d + other.d);
    }

    Quaternion times(Quaternion other) {
        double newA = a * other.a - b * other.b - c * other.c - d * other.d;
        double newB = a * other.b + b * other.a + c * other.d - d * other.c;
        double newC = a * other.c - b * other.d + c * other.a + d * other.b;
        double newD = a * other.d + b * other.c + c * other.b + d * other.a;
        return new Quaternion(newA, newB, newC, newD);
    }

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();

        if (a != 0) result.append(a);
        if (b != 0) result.append(result.length() > 0 && b > 0 ? "+" + b + "i" : b + "i");
        if (c != 0) result.append(result.length() > 0 && c > 0 ? "+" + c + "j" : c + "j");
        if (d != 0) result.append(result.length() > 0 && d > 0 ? "+" + d + "k" : d + "k");

        return result.length() == 0 ? "0" : result.toString();
    }
}

sealed interface BinarySearchTree permits Empty, Node {
    int size();
    boolean contains(String value);
    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0) {
            return new Node(this.value, left.insert(value), right);
        } else {
            return new Node(this.value, left, right.insert(value));
        }
    }

    @Override
    public String toString() {
        String leftStr = left instanceof Empty ? "" : left.toString();
        String rightStr = right instanceof Empty ? "" : right.toString();
        return "(" + leftStr + value + rightStr + ")";
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }

    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// First Then Lower Case function
fun firstThenLowerCase(array: List<String>, predicate: (String) -> Boolean): String? {
    return array.firstOrNull(predicate)?.lowercase()
}

// Say function
data class Say(val phrase: String) {
    fun and(nextPhrase: String): Say {
        return Say("$phrase $nextPhrase")
    }
}
fun say(phrase: String = ""): Say {
    return Say(phrase)
}

// Meaningful Line Count function
@Throws(IOException::class)
fun meaningfulLineCount(filePath: String): Long {
    return BufferedReader(FileReader(filePath)).use { reader ->
        reader.lineSequence().count { it.isNotBlank() }.toLong()
    }
}

// Quaternion data class
data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0) // Corrected I to match complex number convention
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)

    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)
    }

    operator fun times(other: Quaternion): Quaternion {
        return Quaternion(
            a * other.a - b * other.b - c * other.c - d * other.d,
            a * other.b + b * other.a + c * other.d - d * other.c,
            a * other.c - b * other.d + c * other.a + d * other.b,
            a * other.d + b * other.c - c * other.b + d * other.a
        )
    }

    override fun toString(): String {
        val terms = mutableListOf<String>()
        if (a != 0.0) terms.add(a.toString())
        if (b != 0.0) terms.add("${if (b < 0) "" else "+"}${b}i")
        if (c != 0.0) terms.add("${if (c < 0) "" else "+"}${c}j")
        if (d != 0.0) terms.add("${if (d < 0) "" else "+"}${d}k")
        return if (terms.isEmpty()) "0" else terms.joinToString("")
    }
}


// Binary Search Tree interface and implementations
sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean

    data class Node(val value: String, val left: BinarySearchTree = Empty, val right: BinarySearchTree = Empty) : BinarySearchTree {
        override fun size(): Int = 1 + left.size() + right.size()

        override fun contains(value: String): Boolean {
            return when {
                value < this.value -> left.contains(value)
                value > this.value -> right.contains(value)
                else -> true
            }
        }

        override fun toString(): String {
            val leftStr = if (left == Empty) "" else left.toString()
            val rightStr = if (right == Empty) "" else right.toString()
            return "($leftStr$value$rightStr)"
        }
    }

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun toString(): String = ""
    }

    fun insert(value: String): BinarySearchTree {
        return when (this) {
            is Empty -> Node(value)
            is Node -> {
                when {
                    value < this.value -> Node(this.value, left.insert(value), right)
                    value > this.value -> Node(this.value, left, right.insert(value))
                    else -> this // No duplicates allowed
                }
            }
        }
    }
}
